### PR TITLE
[model, perf] refactor: build DeepSeek-V4 sparse indices without a dense mask

### PR DESCRIPTION
**Pull request by FoolPlayer**
> **Stacked on #1013.** That PR restores `patchgen --check` on `main`; this one
> regenerates the same files, so it needs #1013 first to keep the diff readable.
> Review the second commit only.

### What does this PR do?

The packed TileLang path materialized a `[B, 1, S, S + C]` additive mask on
every forward, plus a per-layer compressed block bias, only to filter a compact
candidate list against it. Every constraint that mask encoded is available in
closed form, so this drops it.

The saving scales with `S²`, which is what makes it worth doing at long context:
**0.51 GB of peak memory at 16K**, against 0.03 GB at 4K.

### Checklist Before Starting

- Related: #1013 (base of this stack), #981, #1012
- PR title follows `[{modules}] {type}: {description}` format

### Test

Verified on 4x GB200 with `DeepSeek-V4-Flash-Base-4L`. That checkpoint has
`compress_ratios [0, 0, 4, 128]`, so a single run covers all three layer kinds —
no compressor, CSA and HCA — rather than only the common one.

**Loss is identical to the dense-mask path at every one of 20 steps** at 16K
sequence length. Peak memory drops 0.51 GB at 16K and 0.03 GB at 4K.

New tests in `tests/ops/test_deepseek_v4_kernels.py` prove the two index
builders agree bitwise with the dense-mask filter, including the sentinel and
boundary cases, and an end-to-end test asserts the model actually takes the
mask-free path and hands the kernel the indices it expects.

### API and Usage Example

No API or config change. The path is selected at runtime; there is no flag.

### Design & Code Changes

The mask encoded two things, and both are available directly:

- **Sliding causality and packed-sample isolation** collapse into the single
  test `offset <= position_ids[q]`. `position_ids` restarts at zero on each
  packed sample, so `q - position_ids[q]` is that sample's first token, and any
  larger offset would reach into the previous sample.
- **Compressed entries** are either an indexer selection, which already marks
  misses with `-1`, or the contiguous `[range_start, range_end)` interval the
  block bias was built from.

So `DeepseekV4Model.forward` skips `create_sliding_window_causal_mask` entirely
on that path, and the compressors skip the block bias — which also drops the
per-layer concat and the mask gather in the index filter. The candidate tensor
keeps its width (`sliding_window` + top-k budget) with invalid slots marked
`-1`, so the kernel's topk specialization is unchanged.

**Dropping the mask makes the eager fallback unsafe**, because it would have
nothing left to enforce causality. Two guards cover that: the model only goes
mask-free for bf16 CUDA inputs, mirroring the runtime conditions the TileLang
dispatch itself checks, and the fallback raises rather than silently computing
bidirectional attention when handed validated indices.

The NPU sibling reuses this attention forward verbatim, so its import list and
generated module move in step. The mask-free branch is inert there — it is gated
on the TileLang dispatch and on bf16 CUDA inputs, so NPU keeps taking the dense
mask path.

### Checklist Before Submitting

- Read the [Contribute Guide](https://github.com/ByteDance-Seed/VeOmni/blob/main/CONTRIBUTING.md)
- Applied pre-commit checks (`make quality` and `make check-patchgen` pass)
- Added/updated documentation — n/a, no user-facing surface changes
- Added tests to CI workflow: yes, in `tests/ops/test_deepseek_v4_kernels.py`

### CODE REVIEW

File: veomni/models/transformers/deepseek_v4/deepseek_v4_npu_patch_gen_config.py [hunk 1]
```diff
@@ -138,13 +138,16 @@
 config.add_import(
     "veomni.models.transformers.deepseek_v4.packed_utils",
     names=[
+        "CompressedCandidates",
```
[the remark this answers is not in the document]
Comment (FoolPlayer) [reply veomni/models/transformers/deepseek_v4/deepseek_v4_npu_patch_gen_config.py:143]: Confirmed and fixed in `4df122485`. I reproduced it the same way you did, binding every call site in both generated modules against the callee signature:

```
npu  DeepseekV4HCACompressor.forward <- build_block_bias   *** unexpected keyword ***
npu  DeepseekV4CSACompressor.forward <- build_block_bias   *** unexpected keyword ***
```

Both forks now carry the GPU signature and return `CompressedCandidates`, and the CSA one calls `scatter_topk_block_bias` instead of repeating that scatter inline — which leaves the gradient anchor as the only thing still differing between the two sides.

On the `return_topk_indices` half: that is fixed at the base of the stack (#1013, `b42305b25`), so it no longer arrives here.

On making the class of bug visible to CI: #1013 now carries `[ci] test: bind generated call sites against their callees`, which is essentially the script you quoted, run over every generated module. It reads source rather than importing, so NPU modules are covered on a GPU host. It fails on this commit's parent, and it also caught the `apply_rope` instance in #1015 — worth noting that no CI job could have: these PRs target a branch rather than `main`, so the Ascend workflow never triggers on them.

File: veomni/models/transformers/deepseek_v4/deepseek_v4_gpu_patch_gen_config.py [hunk 2]
```diff
@@ -769,8 +780,12 @@ def deepseek_v4_attention_forward_patched(
         s_aux = self.sinks.narrow(0, head_start, local_num_heads).contiguous()
 
     block_bias = None
-    compressed_topk_indices = None
+    compressed_candidates = None
     use_compact_sparse_indices = veomni_dsa_attention_implementation.value == "tilelang" and past_key_values is None
+    # ``DeepseekV4Model.forward`` withholds the dense mask exactly when the packed
+    # metadata is sufficient to validate candidates on its own, so its absence is
+    # the signal to take the mask-free path and skip every O(S^2) intermediate.
+    mask_free_sparse = use_compact_sparse_indices and attention_mask is None
```
[the remark this answers is not in the document]
Comment (FoolPlayer) [reply veomni/models/transformers/deepseek_v4/deepseek_v4_gpu_patch_gen_config.py]: Agreed on the substance and changed in `1e9ca6c8c`: the decision now travels in `kwargs` next to `packed_sequence_slices`, so the four-condition gate is the single source of truth and the attention layer no longer re-derives anything.

One correction to the specific instance, though. `DeepseekV4PreTrainedModel` sets `_supports_sdpa = False` and `_supports_flash_attn = False`, so `_attn_implementation` cannot be either of those for this model — `from_pretrained` raises before the mask builder is ever consulted. Your measurement of `create_sliding_window_causal_mask` is right, but eager is the only implementation this class accepts and eager always returns a tensor, so the two predicates cannot actually disagree in tree today. The change is hardening, not a live bug fix, and I've said so in the commit message rather than claiming a fix.

What it does buy beyond deduplication is the failure mode: if a layer finds itself unable to build compact indices after the model already dropped the mask, it now raises instead of running attention with no mask at all. As you note, the previous guard sat in `eager_attention_forward` and so was reachable only on that one branch.

Two tests, both of which fail on the previous derivation:

- `test_deepseek_v4_attention_reads_the_decision_not_the_missing_mask` — a layer handed `attention_mask=None` with no packed metadata reaches for the packed builder under the old code, and for the unpacked one now.
- `test_deepseek_v4_attention_refuses_mask_free_it_cannot_serve` — the fallback that would have run unmasked.

File: veomni/models/transformers/deepseek_v4/packed_utils.py [hunk 3]
```diff
@@ -176,6 +193,91 @@ def build_sparse_attention_indices(
...
+
+    * sliding causality and packed-sample isolation collapse into the single test
+      ``offset <= position_ids[q]``. ``position_ids`` restarts at zero on each
+      packed sample, so ``q - position_ids[q]`` is that sample's first token and
+      any larger offset would reach into the previous sample.
+    * compressed entries are either an explicit indexer selection that already
+      marks misses with ``-1``, or a contiguous ``[range_start, range_end)``
+      interval computed from the same metadata that built the block bias.
+
+    Callers must guarantee the padding mask is all ones over the packed length,
```
[the remark this answers is not in the document]
Comment (FoolPlayer) [reply veomni/models/transformers/deepseek_v4/packed_utils.py]: Taken, and enforced rather than downgraded to a comment — you are right that every other precondition in this file raises, so this one reading like a checked contract while not being one was the actual problem.

The check is `attention_mask is None or attention_mask.all()`, added to the model-side gate in `1e9ca6c8c`, so it runs once per forward rather than once per layer, and it sits after the `isinstance(attention_mask, dict)` term so only the None/Tensor cases reach it. The docstring now points at where it is enforced instead of asking the caller to guarantee it. As you say, VeOmni's collator satisfies it by construction, so this is about a foreign collator rather than a live case.

**Comment by FoolPlayer** [comment]
Superseded by #1107, which rebuilds this change on current `main` (after #1012, #1013 and #1016) and fixes the NPU incompatibility raised in review: `DeepseekV4Attention.forward` is shared verbatim with the NPU patch config, so the new `build_block_bias=` keyword reached NPU compressor overrides that did not accept it. Both overrides now take the parameter, and the NPU config's import list gains `build_packed_sparse_attention_indices` / `CompressedCandidates`, without which the generated NPU module referenced an undefined name.
